Vk ext shader object optional layer

gpx1000 opened the pull request:
## Description

This fixes the remaining issues in the newly accepted/merged VK_EXT_shader_object PR.

Commit:
Add VK_EXT_shader_object sample

gpx1000 committed:
add an ability to have optional vulkan layers.  If it's an optional layer, and it's not found, continue to load the application instead of erroring out.
Fix the validation layer discovered issues in shader object.  The render_pass gets created twice (once in the super class prepare and once after the child class recreates the frame buffer.  Destroying the first created renderpass will solve this.).  Also the heightmap_texture and terrain_array_texture both get their sampler's default created and then explicitly recreated.  So adding the destroy sampler just before recreation solves the second and third validation issue.

gpx1000 committed:
Merge remote-tracking branch 'upstream/main'

gpx1000 committed:
Merge branch 'main' into VK_EXT_shader_object_optional_layer

gpx1000 committed:
fix the formatting and copyright breaks.

gpx1000 committed:
get full_screen_exclusive to build.

gpx1000 committed:
clang format locally wants it formatted differently than CI.

gpx1000 committed:
if there's an error in clang-format then save off the diff output so it can be downloaded after the run and applied as a patch.

gpx1000 committed:
attempt to fix clang-format

SaschaWillems reviewed (dismissed):
LGTM

asuessenbach commented on .github/workflows/check.yml:
```diff
@@ -75,7 +75,15 @@ jobs:
+      with:
+        name: diff-output
+        path: patch.txt
+
```
Are these changes related to this PR?

asuessenbach commented on samples/extensions/shader_object/shader_object.cpp:
```diff
@@ -230,6 +230,11 @@ void ShaderObject::setup_framebuffer()
 void ShaderObject::setup_render_pass()
 {
+	// delete existing render pass
+	if (render_pass != VK_NULL_HANDLE)
```
Is `setup_render_pass` called multiple times?
If so, you would need to initialize `render_pass` with `VK_NULL_HANDLE` to make this check work in debug mode. If not, you don't need to destroy the render pass here.

marty-johnson59 replied:
Hi @asuessenbach , per discussion today, can  you open an issue so we don't lose the context?  We're going to merge this one soon w/o capturing this - but want to circle back later

asuessenbach replied:
Sorry, I missed the meeting where this was discussed...
Why should this issue not be resolved right here?

SaschaWillems replied:
The idea is to fix this after this PR has been merged. The PR is kinda urgent, and the line you highlighted is not new to the PR but already present in the base sample.

asuessenbach replied:
@SaschaWillems so it seems, you know about the issue to be resolved later. Would you be so kind and file that issue, instead of me guessing what's meant to be done here?

SaschaWillems replied:
Sorry, I was under the impression that this change was already present beforehand. But as this code is new to this PR you're right to mention this and it indeed needs fixing in this PR.

gpx1000 replied:
I did this change because renderpass was being created by the vulkansample in the framework automatically.  I can take this change out, but then validation layers correctly identifies a render_pass object not getting released.  The rest of the changes are updated to conform with your requests.

SaschaWillems replied:
That makes sense. We should keep that change then :)

asuessenbach reviewed (commented):
This PR works.
But instead of improving the somewhat hacky usage of `VulkanSample::get_validation_layers`, I would completely remove that function, and instead introduce some `VulkanSample::add_instance_layer` (or maybe just `VulkanSample::add_layer`, as there's nothing like a device layer), similar to `VulkanSample::add_instance_extension`, and handle it analogously.

gpx1000 committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into VK_EXT_shader_object_optional_layer

# Conflicts:
#	framework/core/hpp_instance.h
#	framework/core/instance.h
#	framework/hpp_vulkan_sample.cpp
#	framework/hpp_vulkan_sample.h
#	framework/vulkan_sample.cpp
#	samples/extensions/shader_object/README.adoc
#	third_party/volk
#	third_party/vulkan

gpx1000 committed:
Refactor validation layers to use vector instead of map

Related functions and variables have been updated to accommodate the change. This ensures consistency and optimizes the performance for layer searches and validations.

gpx1000 committed:
revert the change in check.yml

gpx1000 committed:
fix the copyright and formating and doxygen problems.

gpx1000 committed:
revert vulkan and volk submodule changes.

gpx1000 committed:
clang format

gpx1000 committed:
update the clang-format

gpx1000 committed:
revert full_screen_exclusive.cpp and hello_triangle.cpp

asuessenbach commented on framework/core/hpp_instance.cpp:
```diff
@@ -74,6 +74,41 @@ static VKAPI_ATTR VkBool32 VKAPI_CALL debug_callback(VkDebugReportFlagsEXT flags
 }
 #endif
 
+bool validate_layers(std::unordered_map<const char *, bool> &required,
```
The `required` layers here actually are not required (they're allowed to be optional), but requested. Maybe rename the argument?

asuessenbach commented on framework/core/hpp_instance.cpp:
```diff
@@ -74,6 +74,41 @@ static VKAPI_ATTR VkBool32 VKAPI_CALL debug_callback(VkDebugReportFlagsEXT flags
+			}
+		}
+
+		if (!found)
```
Maybe we can replace the loop over the `available`s with this check:
`		if (std::none_of(available.begin(), available.end(), [&layer](vk::LayerProperties const &lp) { return lp.layerName == layer.first; }))
`

asuessenbach commented on framework/core/hpp_instance.cpp:
```diff
@@ -74,6 +74,41 @@ static VKAPI_ATTR VkBool32 VKAPI_CALL debug_callback(VkDebugReportFlagsEXT flags
+
 bool validate_layers(const std::vector<const char *>        &required,
                      const std::vector<vk::LayerProperties> &available)
 {
```
In case, you want to change a little more, this version of `validate_layers` could look like this:
```
	auto requiredButNotAvailableIt =
	    std::find_if(required.begin(),
	                 required.end(),
	                 [&available](char const *layer) {
		                 return std::none_of(available.begin(), available.end(), [&layer](vk::LayerProperties const &lp) { return lp.layerName == layer; });
	                 });
	if (requiredButNotAvailableIt != required.end())
	{
		LOGE("Validation Layer {} not found", *requiredButNotAvailableIt);
		return false;
	}
	return true;

```

asuessenbach commented on framework/core/hpp_instance.cpp:
```diff
@@ -299,6 +335,20 @@ HPPInstance::HPPInstance(const std::string                            &applicati
 		throw std::runtime_error("Required validation layers are missing.");
 	}
 
+	std::unordered_map<const char *, bool> layers = (std::unordered_map<const char *, bool>) (requested_layers);
```
No.
Never cast the `const` away.
If you would handle the `requested_layers` the same way as the `required_extensions` (which should be named `requested_extensions`!), you would not even need the new `validate_layers` version.
And, by the way, the `layers` are never used here!

asuessenbach commented on framework/core/instance.cpp:
```diff
@@ -76,6 +76,41 @@ static VKAPI_ATTR VkBool32 VKAPI_CALL debug_callback(VkDebugReportFlagsEXT flags
 #endif
 
+bool validate_layers(std::unordered_map<const char *, bool> &required,
+                     const std::vector<VkLayerProperties>   &available)
```
The same comments as for hpp_instance.cpp hold here.

asuessenbach commented on samples/extensions/shader_object/README.adoc:
```diff
@@ -436,13 +436,15 @@ The layer can be shipped with your application, and it will disable itself if a
 
 The sample framework already has an existing abstraction normally used for enabling the validation layer.
-This sample repurposes this mechanism to instead load the emulation layer.
+This
```
Newline here?

Besides that, is that description still valid? We now have some `add_layer` and `get_layers`. None of them explicitly mentions validation layers.

asuessenbach commented on samples/extensions/shader_object/README.adoc:
```diff
@@ -436,13 +436,15 @@ The layer can be shipped with your application, and it will disable itself if a
 [,CPP]
 ----
-const std::vector<const char *> ShaderObject::get_validation_layers()
+const std::unordered_map<const char *, bool> ShaderObject::get_validation_layers()
```
There's no `ShaderObject::get_validation_layers` anymore.

gpx1000 committed:
Merge branch 'main' into VK_EXT_shader_object_optional_layer

gpx1000 commented on samples/extensions/shader_object/shader_object.cpp:
```diff
@@ -415,6 +416,10 @@ void ShaderObject::load_assets()
+	vkDestroySampler(get_device().get_handle(), heightmap_texture.sampler, nullptr);
+	vkDestroySampler(get_device().get_handle(), terrain_array_textures.sampler, nullptr);
+
 	// Setup a mirroring sampler for the height map
```
looks like the calling load_assets multiple times bug has been fixed.

gpx1000 committed:
Refactor layer validation handling for consistency.

Updated validation layer mechanisms to use `unordered_map` for better clarity and flexibility in distinguishing required and optional layers. Adjusted related structures, methods, and logging to align with the new approach. Simplifies enabling and validation of layers across the framework and samples.

gpx1000 committed:
Fix inconsistent indentation and formatting across files

Aligned indentation and formatting in multiple files to maintain code consistency. Changes include correcting spacing, adjusting alignment in function signatures, and fixing misplaced inline elements where necessary.

gpx1000 committed:
Add get_validation_layers() method and formatting adjustment

Introduce the get_validation_layers() method in ShaderObject to return validation layer information. Additionally, make a minor formatting adjustment to vk::DynamicLoader initialization for readability.
